proposals/0000-reduce-fragmentation.md

```diff
@@ -0,0 +1,1107 @@
+
+Proposed syntax:
+
+```js
+// Either
+css({
+  position: 'absolute',
+  '@media (max-width: 600px)': {
+    position: 'sticky',
+  }
+});
+
+// Or
+css({
+  position: {
+    default: 'absolute',
+    '@media (max-width: 600px)': 'sticky',
+  }
+});
+```
+
+The benefit of the latter is clearer expectations (and greater constraints) about when properties are overridden. For example, it's not as clear what the value for `position` would be in the following:
+
+```js
+css(
+  {
+    position: 'absolute',
+    '@media (max-width: 600px)': {
+      position: 'sticky',
+    }
+  },
+  {
+    position: 'relative',
+  }
+);
+
+// {
+//   position: relative,
+//   '@media (max-width: 600px)': {
+//     position: 'sticky',
+//   }
+// }
+```
+
+Whereas in the next example we can set expectations that a property *and any previous conditions* are completed overridden by any declaration that later modifies the property in any way.
+
+```js
+css(
+  {
+    position: {
+      default: 'absolute',
+      '@media (max-width: 600px)': 'sticky'
+    }
+  },
+  {
+    position: {
+      default: 'relative',
+    }
+  }
+);
+
+// {
+//   position: relative
+// }
+```
+
+Consideration is also required to determine what the final value of a property is when Media Query conditions overlap, i.e., which media features and which conditions (e.g., multiple `width` conditions) take priority over others.
+
+## CSS Properties and Values
+
+Miscellaneous CSS properties and values that should be supported on native.
+
+* [ ] [`backgroundImage`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-image). Add support for setting background images via `url()`. Stretch: support CSS gradients.
+* [ ] [`backgroundOrigin`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-origin).
+* [ ] [`backgroundPosition`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position).
+* [ ] [`backgroundRepeat`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-repeat).
+* [ ] [`backgroundSize`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-size).
+* [ ] [`boxSizing`](https://developer.mozilla.org/en-US/docs/Web/CSS/box-sizing).
+* [ ] [`boxShadow`](https://developer.mozilla.org/en-US/docs/Web/CSS/box-shadow). Add native support for CSS box shadows to replace buggy, iOS-specific `shadow*` styles. Resolves issue (#26110)[https://github.com/facebook/react-native/pull/26110] and [multiple other issues](https://github.com/facebook/react-native/search?q=ios+shadow&type=issues).
+* [ ] [`caretColor`](https://developer.mozilla.org/en-US/docs/Web/CSS/caret-color).
+* [ ] [`clipPath`](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path).
+* [ ] [`display`](https://developer.mozilla.org/en-US/docs/Web/CSS/display) values of `block`, `contents` (resolves proposal [#348](https://github.com/react-native-community/discussions-and-proposals/issues/348)), `inline`, `inline-block`, `inline-flex`.
```

Comment:
If the goal is "the best bits of the web" then I feel like `display: grid` really ought to be supported too.

Reply:
The layout engine is [Yoga](https://github.com/facebook/yoga), and it doesn't support grid (because grid is fairly complicated to build on top of the native platforms).

I feel like grid should be a part of _"Make RN more like Web V2"_ to keep from holding anything else up.

- https://github.com/facebook/yoga/issues/47 (original "support grid" issue from 2015)
- https://github.com/facebook/yoga/issues/867 (2019 updated issue)

Reply:
@corysimmons It doesn't. But then Yoga also doesn't support `block`, `inline` or `float` layout, and they're all included in this proposal. Even Yoga's Flexbox implementation isn't web compatible and will need a rework if the aim is to support web layouts without modification.

Grid is complicated, but it's not really much more complicated than flexbox. I link to an [existing native code implementation](https://github.com/DioxusLabs/taffy) that could be either be used alongside Yoga or ported into Yoga in the issue you mention (https://github.com/facebook/yoga/issues/867) (disclaimer: I wrote said implementation - but this is not me just shilling my own library - I wrote it in large part *because* I want to see it in React Native)

Is Grid just considered less essential? It's one of the most highly requested feature by React Native end users (me being one them!). Surely it at least ought to be considered higher priority than `float`, no?

Reply:
Yeah, offline I had advocated for "grid" before "block" or "inline", and I do think it would be likely to come first (though I would also believe there is more web code using the latter if porting existing web code is the main draw).

It is exciting that Taffy has an implementation of Grid as well, though I want to set expectations that we are not planning to sequence work to port grid in the short term. I don't think we have solved safely innovating on Yoga yet, nor do I think we have really solved problems which would let us smoothly transition RN to any alternative layout engines. Both are at the scale of hundreds of thousand of surfaces, and billions of users, and even small changes in behavior of performance characteristics can effectively take out major portions of the industry.

To create better clarity there is a plan I am going to share out for concrete Yoga + RN work we are taking on over the next couple of months, as a more concrete articulation of the vision and accompanying discussion shared late last year. The majority of it is around shaping Yoga/RN to start letting us make conformance changes more confidently, along with getting a new OSS release of Yoga out.

Reply:
@NickGerleman I guess I'm pretty skeptical of what appears to be the proposed approach. Which as I understand it is something like the following:

> Incrementally evolve the existing algorithm towards standards compliance using fine-grained feature flags to alter behaviour in each place where the standards-compliant behaviour differs from the legacy behaviour while keeping a shared implementation for parts where the behaviour is the same.

It seems to me that by taking such an approach you end up with what is essentially two different algorithms sharing an implementation and that this is good nobody:

- **It's not good for users of React Native who's apps are in maintenance mode and need strict backwards compatibility**, because it means that almost any change to Yoga, even if it aimed at the new "web compatible" version of the algorithm, is likely to touch codepaths that their app uses. This greatly increases the chance of a regression / backwards incompatible change being introduced.

- **It's not good for users of React Native who want web-compatible layout** because it means development will be much slower (see point 3) Which means that either such an implementation will not be available at all for some time to come (years?), or that there will ongoing breaking changes once they adopt the algorithm. Which may well not tolerable even for people who willing to do a one-time pass over their code to switch to a new algorithm.

- **It's not good for maintainers of / contributors to Yoga** as it will be much harder to make sense of each algorithm's flow when the implementations of each are intertwined with each other. And also because it meants that every change has to be checked for compatibility with the legacy algorithm (even if it is only intended to affect the new one), for which Yoga does not have a good test suite and testing is slow. This also effectively all-but-excludes OSS contributors from contributing algorithm changes to Yoga as they will not have access to the testing infrastructure that Meta employees are using the to test for compatibility with legacy apps.

I also think you might be underestimating the amount of changes that would be required to make Yoga's current implementation standards compliant. Based on my experience making similar changes to Taffy (which is derived from an earlier version of Yoga, and even after the changes made so far is still not entirely standards compliant), I would not be at all surprised if a shared implementation of RN's current algorthm and a standards-compliant flexbox algorithm ended up with more code feature flagged than shared. I would expect the tweaks to be small, but large in number and widespread throughout the algorithm.

---

My counter-proposal would be something like the following:

1. Adapt the current implementation such that rather than recursively calling itself when it needs to size or layout a child node, it calls into a `layout_child` function that it takes as a function pointer parameter (or some such similar arrangement that allows the caller to provide a function to call to size child nodes). This will allow either Yoga itself or the embedder of Yoga to take back control flow between each node in the hierarchy and enables the possibility of switching between different algorithms/implementatons for each node.

2. Freeze the current algorithm as it is. Aside from things like minor changes for compatibility with new compiler or similar, this code doesn't get touched from this point onwards.

3. Implement a new standards-compliant algorithm. This could either be a copy of the current algorithm adapted to be more standards compliant (but free from the burden of backwards compatibility). Or it could be a port of an existing implementation that is already (more) standards compliant. Or it could be directly calling into an existing implementation that is already (more) standards compliant.

4. In terms of releasing the new algorithm to RN users: it should be done in an opt-in manner on a per-node basis. I would suggest something like a new `Div` component (or `Flex`/`Row`/`Column` components), leaving `View` to continue using the legacy algorithm indefinitely, but there are a few different ways that you could handle this. I also think some effort should be put into tightening up the standards-compliance of the new implementation before release (perhaps porting the WPT test suite) so as to largely avoid the need for further breaking changes (although it could potentially be released as a preview version first, similarly to how to Hermes was introduced).

Step 1 would need to be done carefully in order to avoid breaking backwards compatibility, but no more carefully than *every single change* would need to be made if you go down the shared implementation route. But once you've made that changes then *most* algorithm changes to the new algorithm wouldn't need nearly as much scrutiny as legacy users simply wouldn't use that codepath at all. And as far as I can see, such a change would need to be made anyway if you want to support mutliple layout algorithms (even ones as simple as `display: block`).

Reply:
> It seems to me that by taking such an approach you end up with what is essentially two different algorithms sharing an implementation and that this is good nobody:

We end up with different branches in Yoga, for each case we find a bug. And a way in Yoga to specify whether we want conformance or compatibility, to let users of Yoga control the breaks. StrictLayout is the way we expose this to React Native, and provides immediate value to cross-users with web, but the plan is to encourage migration via warnings of relying on inconsistent behavior. This does add some debt to Yoga, when we discover a behavior is relied upon. Though we have a framework for deciding what changes are encountered rarely enough to make the break.

---

> I also think you might be underestimating the amount of changes that would be required to make Yoga's current implementation standards compliant.

Followed by 

> Implement a new standards-compliant algorithm. This could either be a copy of the current algorithm adapted to be more standards compliant (but free from the burden of backwards compatibility).

We do though have a queue already of conformance issues we know about, and so far the implementation has not been the tricky bit. Because we have the existing queue, I am seeing WPT exploration as a lower priority, until we start landing the fixes. Though I would consider them to be a requirement for adding new features/algorithms.

---

> Adapt the current implementation such that rather than recursively calling itself when it needs to size or layout a child node, it calls into a layout_child

The design behind `LayoutableShadowNode` has this sort of scenario in mind in Fabric. This is the sort of thing I think we could do if we wanted escape hatches to different layout systems per-node. Historically I think this has come up with if we wanted RN to have a lower level concept like `HStack` or `VStack`, since we have heard from other teams this can be implemented more performantly as its own primitive instead of directly on top of flexbox.

---

> It's not good for users of React Native who's apps are in maintenance mode and need strict backwards compatibility, because it means that almost any change to Yoga, even if it aimed at the new "web compatible" version of the algorithm, is likely to touch codepaths that their app uses.

> Freeze the current algorithm as it is.

I have been exploring the strategy of freezing the algorithm, but am not sure we should follow it, as we will need to maintain both new and old Yoga then. As they diverge, this becomes harder, which is why I have so far been more interested in instead allowing Yoga to switch its own algorithms or branch for bugfixes.

So we do instead continue to change Yoga, with an intentionality to fixes. Though specific users of Yoga may continue to use older versions if they are hesitant to use a layout engine which has any active work being done.

---
 
> In terms of releasing the new algorithm to RN users: it should be done in an opt-in manner on a per-node basis. 

Is the exact strategy we are taking with StrictLayout. This is just controlled by a per-Node Yoga config, instead of a new layout engine.  

---

In other threads I have seen your advocacy of replacing Yoga with Taffy and adding a per-node flag to switch to the other engine, and I think maybe this is some of the motivation here? I have seen Taffy as a healthy project, and have been impressed by the work happening, but a lot of this has been coming with a suggestion that Taffy is more conformant, feature-rich, and conformant as Yoga. Apart from some gaps I think you are aware of (e.g. RTL support), I think we would need a lot more evidence, to make these sorts of assertions.

Reply:
> I have been exploring the strategy of freezing the algorithm, but am not sure we should follow it, as we will need to maintain both new and old Yoga then. As they diverge, this becomes harder, which is why I have so far been more interested in instead allowing Yoga to switch its own algorithms or branch for bugfixes.

As I see it, the old algorithm shouldn't require much maintenance. Only occasional minor fixes to ensure it continues to compile on newer compilers (which given C++'s generally good backwards compatibility story should be minimal). There should be no need to make any algorithm changes to the frozen algorithm. If users want changes or fixes then they can migrate to the new algorithm. The algorithms diverging over time therefore doesn't create any extra work under a model where it is frozen and ongoing changes happen in a copy, as you're only making meaningful changes to the new algorithm.

However, if the old algorithm and the new algorithm share code then I think *that* will make the code harder and harder to maintain as the algorithms diverge. Because:

1. You end up with branches in the code all over the place, making it hard to read the code and mentally walk through it's execution

2. Any changes made to shared codepaths have to be validated for correctness against two different standards. One of which (the current behaviour) we do not a test suite for.

---

Regarding the latter point. You say that:

> we have a framework for deciding what changes are encountered rarely enough to then only include new behavior.

However my understanding is that this process amounts to something like "make the change in a branch of the meta monorepo and then run CI across everything affected". And while this is a useful tool to have it is:

1. Not available to potential contributors outside of Meta

2. A relatively slow process to run.

It does not seem like the sort of tool like that could be run iteratively after making a small code change to quickly gain feedback on whether that change has fixed and/or regressed correct behaviour. And I see that as a big problem:

Suppose I (or some other OSS contributor) contributes a PR that updates Yoga to be more standards conformant. They make the changes and include generated tests to show that the changes are correct according to web standards. They then submit their PR but it gets rejected because it breaks the legacy algorithm. Their PR is now effectively blocked. And what's more, that contributor is not in a position to fix it because they are unable to verify if their changes are correct other than via a process of pushing changes, pinging a Meta employee to test their changes and report back. You'd likely end up with iteration cycles measured in days if not weeks. That's not a practical way to develop, and I certainly can't see myself contributing in such an environment.

If you split the algorithms at a high level then you still have this same problem for the frozen algorithm. But as changes to that algorithm would be both rare and minor this is much less of an issue. Anyone is empowered to make major changes to the new version of the algorithm, and has the tools available to verify the correctness of those changes.

IMO this would have a huge impact on development velocity.

---

Regarding how changes are exposed in React Native: I think we are largely in agreement.

---

Regarding Taffy:

- It's true that I am motivated to have it used. I spent a long time working on implementing CSS Grid in Taffy and improving the performance and spec conformance of it's Flexbox implementation, and I would love to see that work being made use of. Especially in React Native, as I am a user of RN myself, and would therefore get to benefit personally from improvements to it's layout capabilities.
- *However*, I also genuinely believe that a split algorithm approach is best regardless of whether Taffy is used. Taffy seems like the quickest way to get access to new algorithms (CSS Grid and standards conformant Flexbox) in Yoga, as you can call a function from a library rather than having to write and validate the entire algorithm. But if using Taffy is considered off the table (perhaps because it's in Rust, or because the code is not owned by Meta), I would still advocate for writing the new algorithm separately from the existing one for the reasons explained above.

Regarding Taffy's performance:

- We have benchmarks showing that we're comparable in performance to Yoga (see https://github.com/DioxusLabs/taffy#benchmarks-vs-yoga). We've actually improved performance a little since those were measured. In general, Taffy tends to faster in "deep" benchmarks (few nodes per container but more levels of nodes) and Yoga faster in "wide" benchmarks (many nodes per container but fewer levels), but there isn't much in it either way.

Regarding Taffy's spec conformance:

- Taffy is a fork of [Stretch](https://github.com/vislyhq/stretch) which was created by Emil Sjölander who was also a key contributor to Yoga between 2016 and 2018. He has [claimed](https://news.ycombinator.com/item?id=19781441) that Stretch was more web compatible than Yoga and was in a position to know.
- Certainly it is the case that Taffy, via a combination of the changes made in Stretch and further changes to Taffy itself, no longer suffers from the same deviations from the standard that Yoga is known to have around min/max sizes/constraints and aspect ratio (documented  [here](https://github.com/facebook/yoga/blob/b17f08ed4915f24b794a7f2ffdc5cd0acfaf88e1/yoga/Yoga.cpp#L2663) and [here](https://github.com/facebook/yoga/blob/b17f08ed4915f24b794a7f2ffdc5cd0acfaf88e1/yoga/Yoga.cpp#L781).
- Taffy passes a superset of Yoga's test suite. This includes all tests from both Yoga's `main` and the imported tests from FlexLayout (https://github.com/facebook/yoga/pull/1194) which amounts to 264 tests + an additional 177 tests of our own. I'm not sure how Yoga would do on those tests, but that might be an interesting experiment to do.
- That said, Taffy certainly isn't 100% spec compliant either. There are 3 known issues I am aware of with Taffy:
   - No RTL support
   - No support for `overflow` (although it's affect on layout is minor)
   - Taffy incorrectly resolves percentages against min-heights where it should refuse to.
   
Personally I think Taffy would represent an easier starting point for achieving spec compliant Flexbox (and better than starting from scratch for achieving CSS Grid!). But that may be influenced my better familiarity with Taffy, and also that my Rust is stronger than my C++.

---

Regarding "supporting multiple layout engines":

>  I am not wanting to put us in a position where we support multiple layout engines long term.

That seems reasonable. And maybe that's a good reason to have in-tree implementations of the new web compatible algorithms. But from the way you are talking about things upthread, development of such algorithms is not considered a priority and there are no immediate plans to carry out development of that kind. What is the point of an RFC like this if it isn't actually going to be implemented?

It seems to me that either depending on another library or porting it's algorithms into Yoga would be a good way to get a headstart on such development and make it practical to implement these things in a reasonable timeframe. I see no reason, for example, why React Native shouldn't have both Standards Compliant Flexbox and CSS Grid available by the end of the calendar year (in preview at least).

Reply:
@nicoburns I'm sympathetic to your perspective and would like React Native to be able to benefit from your experience, expertise, and work. I like the idea of getting to this "end state" API sooner rather than later, if only to (in)validate its worth. And we have limited engineering time at Meta. Looking through the Taffy issues, etc., I wonder what your view on how we could support flow layout (display block, inline, inline-block...even `float`) on native? As much as people want CSS Grid too, at the moment we can't even render a `<div>` with the same default layout as a browser would, but that's what a lot of existing React DOM code relies on.

Reply:
@necolas `display: block` should be quite straightforward (which would at least solve the simplest cases of `<div>`s using the default (block) layout)

---

However, `display: inline`, `display: inline-block` and `float` would be much more complicated. This is because these layout modes interact closely with text layout. For this reason, they couldn't really be implemented in either Yoga or Taffy.

In the Rust GUI world the goto library for text layout is [cosmic-text](https://github.com/pop-os/cosmic-text), but cosmic-text assumes that you will be do low-level rendering and handling things like text-selection using raw input events (i.e. more of a Flutter approach than a React Native approach to rendering).

I believe React Native uses the platform's native text-rendering capabilities. So I think it would be limited to whatever is supported by those. I'm not familiar with iOS/Android native text view support. So I'm not sure how much is even possible. I know that React Native supports images in Text components, so some kind of content-inline-with-text must be possible. But I also know that View inside Text is not supported / is buggy. And it's potentially a lot more complex when you consider things like the fact that an `inline-block` box might itself contain text, and that text selection should work seamless across the boundary!

So I think that most of an implementation of `inline/inline-block/float` would need to be in the platform-specific code rather than a layout library like Yoga/Taffy. Perhaps support for laying subtrees could be improved, although I suspect this could be made to work with Yoga's existing measure_func support.

---

If the goal is to render "typical" web content as-is, there are a couple of other features which I think merit consideration:

- `position: static`: i.e. the ability to position `position: absolute` relative to a non-direct-parent ancestor node. Neither Taffy nor Yoga implement this currently (Yoga has support for specifying this style, but as far as I can tell it doesn't actually use it to alter the layout!). Quite a lot of web layouts use this IME (mine do anyway). This shouldn't be too hard to add. The main issues would be threading a context down with the correct offsets, and perhaps adding a way to record which ancestor the node ends up being relative too (although this may not be necessary). There may also be non-layout related complexities as IIRC on web it's possible for a `position: absolute` node to "escape" a scroll view and not scroll despite being a child of that scroll view.

- Table layout. I would definitely consider this in scope for Taffy. However, it's pretty complex (at least non-fixed table layout) and I suspect it would probably be at least a 6 month project for someone to implement it.

---

Finally, it seems worth mentioning that there is this library which implements some of this on top of the existing React Native, but it's unsurprisingly quite limited https://meliorence.github.io/react-native-render-html/docs/reinvent-the-wheel

Reply:
Good to know. I think supporting `display:block` in some form is the most significant part of unblocking initial compatibility for element layout. Are you saying you think that's relatively straight-forward to add to Yoga too?

I think RN can already approximate inline text layout, static positioning, and even flexbox well enough to unblock some next steps related to DOM compat...if we set some restrictions in place for now (e.g., no `inline` support for block elements). My interest in primarily in getting `div` rendering as you'd expect on the web; not in changing the default layout used by `View` in existing surfaces.

Reply:
> Are you saying you think that's relatively straight-forward to add to Yoga too?

Yes, I think so. I think you'd need to implement changes similar to https://github.com/DioxusLabs/taffy/pull/246 which factors out child sizing so that it isn't assumed that a child is `display: flex`.

The block layout algorithm itself is pretty simple (https://www.w3.org/TR/CSS22/visuren.html#block-formatting), although it should be noted that there is a decent amount of complexity introduced by margin collapsing (https://www.w3.org/TR/CSS22/box.html#collapsing-margins), in particular because margins can collapse between parents and children rather than just between siblings. Which I think means that the block algorithm may need to recurse arbitrarily far down the tree whereas the flexbox algorithm only need to look at the styles of direct children.

> My interest in primarily in getting div rendering as you'd expect on the web; not in changing the default layout used by View in existing surfaces.

If you're happy with how Text works and don't need `display: inline-block` then that seems achievable to me :)

Reply:
Well there's always room for improvement but I'm hopeful that this would be enough to motivate further changes in the future! Thanks for taking the time to help me learn a bit more about all this

Reply:
@necolas Taffy has just landed (https://github.com/DioxusLabs/taffy/pull/474) block layout support. The implementation (https://github.com/DioxusLabs/taffy/blob/main/src/compute/block.rs) was actually suprisingly straightforward (sitting at ~650 LoC, with a couple of relatively minor tweaks to the interface we use to allow parents to measure the size of their children also being necessary in order to accomodate margin collapsing). And we have just over 200 (yoga-compatible) test fixtures to go with it.

However, I have another question for you around supporting `<div>`s with default styles: would you consider `box-sizing: content-box` support necesary? Because Yoga and Taffy both only support `box-sizing: border-box` currently. I figure that most of the web is resetting `box-sizing` to `border-box` but technically it isn't the default.

Reply:
I think `context-box` is another "improvement" rather than blocker. For context, the way I've implemented this API is to create a new library that exports "strict" DOM components. A cross-platform component looks like this:

```js
import { html } from 'react-strict-dom';
import { stylex } from '@stylexjs/stylex';

const styles = stylex.create({
  root: {}
});

function MyComponent(props) {
  return (
    <html.div style={styles.root} {...props} />
  );
}
```

The `html` components are as designed in this spec, where `style` is a special prop for all platforms - it's more like the React Native one that then React DOM one. The `html` components on web come with a built-in style reset that normalizes the layout across web and native, including setting `box-sizing:border-box`. There's no way around that to get this off the ground, so we'll probably be "stuck" with `border-box` as the default long-term (which I expect most web products will be ok with, including if they migrate from `react-dom`). But eventually there may be benefits to including support for `content-box` too.

At least, that's my thinking right now.

Reply:
Following up on my previous comment. Internally we've had engineers surface rendering issues that were traced back to elements being styled with `context-box` as the assumed `box-sizing`, and switching those elements to `border-box` was an unexpected event that required style modifications to avoid regressions. So I think in the short-term, it may be an acceptable cost for early adopters rendering existing React DOM UI on native to have to experience regressions like this and do manual adjustments. But if you're trying to "migrate" larger amounts of existing React DOM code to render on native, the native renderer will eventually need the default <div>/<span> layouts to be accurate. I think this is especially valuable for UI because you often only see visual regressions once they've already broken something in production, unless your application has extensive screenshot testing (which most don't/can't).